src/decomposition.jl

```diff
@@ -8,15 +8,51 @@ or the constraint is located.
 This method is called by the `JuMP.optimize!` hook.
 """
 function register_decomposition(model::JuMP.Model)
-    # Link to the tree
-    tree = gettree(model)
-    tree === nothing && return
-    for (_, jump_obj) in model.obj_dict
-        _annotate_elements!(model, jump_obj, tree)
+    if model.ext[:automatic_dantzig_wolfe] != inaktive
+        register_automatic_dantzig_wolfe(model)
+    else
+        tree = gettree(model)
+        tree === nothing && return
+        for (_, jump_obj) in model.obj_dict
+            _annotate_elements!(model, jump_obj, tree)
+        end
     end
     return
 end
 
+function register_automatic_dantzig_wolfe(model::JuMP.Model)
+    tree = gettree(model)
+    # Annotate master constraints
+    decomposition_structure = model.ext[:decomposition_structure]
+    _annotate_elements!(model, collect(decomposition_structure.master_constraints), tree)
+    # Annotate variables in blocks
+    variables_in_block = Set{MOI.VariableIndex}()
+    annotated_variables = Set{MOI.VariableIndex}()
+    axisids = Vector{AxisId}()
+    virtual_axis = BlockDecomposition.Axis(1:length(decomposition_structure.blocks))
+    for i in virtual_axis
+        empty!(variables_in_block)
+        empty!(axisids)
+        push!(axisids, i)
+        # Annotate constraints in one block (and variables contained in these)  with the same annotation
+        ann = _getannotation(tree, axisids)
+        for constraintref in decomposition_structure.blocks[i]
+            setannotation!(model, constraintref, ann)
+            union!(
+                variables_in_block,
+                model.ext[:decomposition_structure].model_description.constraints_to_variables[constraintref]
+            )
+        end
+        for v in variables_in_block
+            setannotation!(model, JuMP.VariableRef(model, v), ann)
+        end
+        union!(annotated_variables, variables_in_block)
+    end
+    # Annotate all other variables
+    all_variables = MathOptInterface.get(model, MathOptInterface.ListOfVariableIndices())
+    _annotate_elements!(model, [JuMP.VariableRef(model, v) for v in collect(setdiff(all_variables, annotated_variables))], tree)
```

Comment:
This line is difficult to read. I think you should consider to create the array before and give to this array a name that describe its content.

Reply:
I agree, thanks!